### PR TITLE
feat(tasks): per-task event correlation + task-detail read API (#185)

### LINKED ISSUE #185: feat(tasks): per-task event correlation + task-detail read API
## Problem

The Task Detail screen (and the lineage + timeline it renders) needs a **per-task event stream**, and nothing assembles one today:

- Events are journaled **company-scoped**. `CompanyEvent::TaskDispatched { task_id }` carries the id, but the events a dispatch then produces (tool calls, agent replies, waits) are **not tagged with `task_id`**, so they cannot be filtered back to their task.
- There is **no task-completion event** — the 14 `CompanyEvent` variants (`src/ports/types.rs`) have nothing for "this task finished", so a lineage rollup / timeline terminal has no anchor.
- There is **no read route** that assembles a per-task view. `GET …/tasks` lists board cards only; the timeline primitive (`AgentReply.steps: Vec<TurnStep>`, scrubbed, persisted, rehydrated) is keyed to a chat bubble, not a task.

## Proposal

Backend foundation for the Task Detail screen (#TASK_DETAIL_EPIC):

1. **Thread `task_id` onto dispatch-produced events** so a task's tool calls / replies / waits can be filtered out of the journal. Additive + `skip_serializing_if` so every already-persisted event round-trips byte-identical (same pattern as `OperatorMessage.by`/`chat`).
2. **Add `CompanyEvent::DeskTaskCompleted { task_id, desk, output }`** — the completion anchor the timeline terminal and lineage rollup need. (#171's done-transition can consume it; this issue only adds the event, not the column write.)
3. **`GET …/tasks/{id}`** read route that assembles: header fields, the timeline (journal filtered by `task_id`, reusing the scrubbed `TurnStep` shape — never raw args/output/ids), lineage (parent/children), and the approvals trail (from the `runtime/journal.rs` immutable ledger).

## Alternatives

- Derive the timeline purely client-side from chat history — rejected: a board-native card has no chat thread, and it loses task correlation entirely.
- One giant per-company event feed the frontend filters — rejected: leaks cross-task events, no server-side lineage.

## Scope / Related

Medium (backend only). Prereq for the Task Detail screen. Refs #176 (brain-agnostic seam / PR #178), #171 (done terminal consumes the completion event), #183 (MVP task board).

### PR DESCRIPTION
**Pull request by M3gA-Mind**
## Summary

Backend read foundation for the Task Detail screen (#185), in the three parts the issue asks for:

1. **Thread `task_id` onto dispatch-produced events.** `AgentReply` and `McpCallFailed` gain an optional `task_id`, stamped by the harness when the producing turn ran inside a `TaskDispatched` cycle. The journal is company-scoped, so without this a dispatch's replies and failed tool calls are indistinguishable from every other desk reply in the log — which is exactly why a per-task filter could not be built before.
2. **Add `CompanyEvent::DeskTaskCompleted`.** Journaled after the card's landing column is persisted, so it always records a completed run. Only the event is added here; #171's done-transition can consume it.
3. **`GET …/tasks/{task_id}`.** Assembles the header, the timeline, the lineage, and the approvals trail into one call.

Lineage gets its own `TaskRecord.parent_task_id`.

## API Or Behavior Changes

**New route** — `GET …/tasks/{task_id}` under both scope forms. 404s on an unknown id, matching `PATCH` / `DELETE`.

```jsonc
{
  "task": { "id": "t-1", "title": "Ship it", "column": "in_review", "parentTaskId": "t-parent", ... },
  "timeline": [
    { "seq": 12, "atMillis": 1, "kind": "dispatched", "label": "Dispatched" },
    { "seq": 13, "atMillis": 2, "kind": "reply",      "label": "Reply from ceo", "detail": "on it" },
    { "seq": 14, "atMillis": 3, "kind": "completed",  "label": "Finished on ceo → in_review", "detail": "shipped" }
  ],
  "lineage": {
    "parent":   { "id": "t-parent", "title": "Parent", "column": "in_review" },
    "children": [ { "id": "t-child", "title": "Subtask", "column": "backlog" } ]
  }
}
```

**New event variant** — `DeskTaskCompleted { task_id, desk, output, column }`.

`column` is beyond the shape the issue sketched, deliberately. "Completed" here means *the run stopped*, not *it succeeded*: a cancelled, paused, or failed dispatch emits one too. Without the landing column a timeline terminal cannot tell those apart, and would have to re-derive intent by parsing `output`.

**New optional fields** — `AgentReply.task_id`, `McpCallFailed.task_id`, `TaskRecord.parent_task_id`, plus `parentTaskId` accepted on `POST /tasks` and `PATCH /tasks/{id}` and returned on the card.

All three are additive (`#[serde(default, skip_serializing_if = …)]`), so every already-persisted event and board row loads unchanged, and an untagged record serializes **byte-for-byte** as it did before the field existed. No stored log needs migrating and the cross-backend export/import round-trip is unaffected. Pinned by `task_id_correlation_is_additive_and_omitted_when_absent`, which asserts the exact legacy JSON string.

**Two behavior fixes** that fell out of the wiring — see *Notable decisions*.

**Not changed:** every existing route's response shape, the SSE projection's deny-by-default stance (the new variant projects domain fields only; the operator identity on an approval is dropped as elsewhere), and desk chat history.

## Tests

- [x] `cargo fmt --all -- --check`
- [x] `cargo clippy --all-targets -- -D warnings`
- [x] `cargo build --all-targets`
- [x] `cargo test`

All four run locally on this branch, all exit 0. `cargo test` → **762 passed, 0 failed, 1 ignored** (+6 in the second target).

Additionally, because `src/harness/` is behind the `openhuman` feature and the default build never compiles it:

- `cargo check --lib --tests --features openhuman` — clean.

That extra check earned its place. It caught a missing `parent_task_id` in `spawn_task`, and then a stale call signature that existed **only in feature-gated test code** — which `cargo check --lib` also misses, because it does not compile tests. Both would otherwise have reached CI green-looking from the default build.

New tests:

| Test | Pins |
| --- | --- |
| `task_detail_assembles_timeline_and_lineage` | The filter, positively **and** negatively: a reply tagged to this task is admitted while an untagged chat reply and a reply tagged to a *different* task are both excluded. Plus parent/children and the 404. |
| `inflight_read_is_not_shadowed_by_task_detail` | `GET …/tasks/inflight` still reaches the operator strip — see *Notable decisions*. |
| `task_id_correlation_is_additive_and_omitted_when_absent` | Legacy events load; an untagged event's serialized form is byte-identical to the pre-field string. |
| `desk_task_completed_round_trips` | The new variant's wire shape. |

## Documentation

`docs/spec/runtime/ports.md` — added `DeskTaskCompleted` to the `CompanyEvent` list and a new **Per-task event correlation** section covering why the correlation key is needed, the additive/no-migration contract, and why `parent_task_id` exists separately from `origin_chat_id`.

## Notable decisions

**Lineage needed a new field.** `origin_chat_id` names a *conversation* — shared by every sibling spawned in that thread, and absent entirely on a board-native card — so it cannot express the task-to-task edge. `parent_task_id` does.

Where it is written: `run_delegation` is only reached from an orchestrator **chat** turn (`run_task` does not drain the delegation queue, and a dispatched card's responder is a desk member, which carries no delegation tools — "no sub-agent re-delegation in v1"). So no task is ever in scope there to *be* the parent today. Rather than ship a field nothing can populate, lineage is written through the task API's `parentTaskId`; the delegation site is commented as where it will be stamped once task turns gain delegation tools.

**Two latent bugs fixed as fallout:**

- `run_task` never drained the MCP failure queue. A dispatch's failures were left for whichever **chat** turn drained next — mis-attributed to an unrelated bubble, and leaving the dispatch's own timeline silent about the very calls that broke. It now drains in place, tagged with the card.
- Giving `{task_id}` a GET for the first time makes it genuinely overlap the static `…/tasks/inflight` the operator strip reads. The static-segment-wins ordering went from belt-and-braces to load-bearing. A card can never be named `inflight`, so a regression here would silently 404 the strip with nothing else failing — hence the dedicated test and the rewritten comment at the route site.

**A behavior change I introduced and backed out.** The task reply was first journaled against the card's `origin_chat_id`, which would have injected it into that desk's chat history via `chat_history::owns` — and duplicated the live post-back bubble. It is now keyed to the **card id**, which matches no desk, so the record stays what it is: timeline material reachable only through `task_id`. A read foundation should not change chat behavior.

**Approvals are correlated by run window**, not by id — parked effects carry no task id, so entries are labelled as window-scoped rather than presented as a false-precision per-task link.

## Related

- Closes #185
- Refs #176 (brain-agnostic seam), #171 (done terminal consumes `DeskTaskCompleted`), #183 (MVP task board)

## Summary by CodeRabbit

- **New Features**
  - Added task detail views with per-task event timelines, completion events, and task lineage (including parent/child relationships).
  - Introduced optional task correlation on relevant reply and tool-failure events, emitted only for dispatch-produced items.
  - Extended delegated task flows to improve completion wording and reporting.

- **Bug Fixes**
  - Fixed tasks routing to avoid endpoint read collisions.
  - Improved task timeline retrieval to use scoped, paged traversal with accurate windowing.

- **Documentation**
  - Documented event correlation rules, additive/backward-compatible fields, and lineage reconstruction.

- **Tests**
  - Updated/added coverage for serialization, SSE projections, timeline scoping, and lineage validation.

### CODE REVIEW

File: src/harness/brain.rs [hunk 1]
```diff
@@ -209,13 +205,68 @@ impl HarnessBrain {
...
+                    &self.record.id,
+                    CompanyEvent::DeskTaskCompleted {
+                        task_id: card.id.clone(),
+                        desk: responder.clone(),
+                        output: result_text,
+                        column: card.column.clone(),
+                    },
+                )
+                .await?;
+        }
```
[the remark this answers is not in the document]
Comment (M3gA-Mind) [reply src/harness/brain.rs:315]: Fixed in 81e636b — agreed, and the failure mode is worse than "an event is missing": the card is already committed at that point, so a journal error abandoned the `DeskTaskCompleted` anchor *and* the #151 post-back, leaving a timeline stuck "still running" for a card the board already shows in `in_review`.

All three writes are now log-and-continue, matching `chat_and_emit` and the other journal-after-persist sites.

One deliberate difference from the suggested diff: the terminal anchor is attempted **even if the reply write failed**, rather than short-circuiting. Dropping the anchor is strictly worse than dropping a reply — the anchor is what closes a timeline window, so losing it also un-scopes the window-correlated approvals arm.

File: src/server/ops/tasks.rs [hunk 2]
```diff
@@ -176,6 +208,224 @@ async fn delete_task(
...
+
+    let stored = company
+        .runtime
+        .events()
+        .read_from(company.id(), EventSeq::new(0), usize::MAX)
+        .await?;
+
+    let mut timeline = Vec::new();
+    let mut window_open = false;
+    for ev in &stored {
```
[the remark this answers is not in the document]
Comment (M3gA-Mind) [reply src/server/ops/tasks.rs]: Fixed in 81e636b, but with one deliberate departure from the suggestion.

`task_timeline` now reads in bounded pages (`TIMELINE_PAGE = 512`) instead of `usize::MAX`, so the log is still traversed but never all resident at once.

**I did not take the early-exit half of the suggestion**, and I think it would be a correctness regression: a card can be re-dispatched — moved back to `in_progress` after review — which opens a *second* dispatch → completion cycle later in the same log. Stopping at the first `DeskTaskCompleted` would silently truncate every run after the first, which is a worse bug than the cost it saves, and one that would only show up for exactly the long-lived tasks the tab is most useful for.

Agreed that paging bounds memory, not traversal. A stored per-task dispatch offset is the real fix for the scan cost; that is a schema addition and belongs to the epic rather than this read foundation — noted in the function docs so the next person does not re-derive it.

**Comment by M3gA-Mind** [comment]
Addressed the outside-diff findings and the nitpick in 81e636b — replying here since those three had no inline thread to answer on.

**`src/server/ops/tasks.rs` — no validation on `parent_task_id` (Major).** Valid, fixed. `create_task` and `patch_task` now validate at the write boundary: self-parenting, a parent naming no existing card, and cycles are all rejected with a 400.

I went further than the suggested guard and included **cycle** detection, since the finding named it. `patch_task` keeps the board `Vec` (no longer consumed by `into_iter`) and walks ancestors up from the proposed parent; reaching the child means the edge would close a loop. The walk carries a visited set so it terminates even against a board that already contains a loop from before this validation existed. On create the child has no id on the board yet, so the cycle check is skipped as unreachable rather than faked.

You are right that nothing hangs today — `task_detail` walks a single level — but a persisted cycle is a latent trap for any future consumer that does walk the chain (a rollup, a breadcrumb), and the write boundary is the cheap place to keep the lineage a forest. Covered by `parent_task_id_rejects_self_unknown_and_cycles`, including the A → B → A case.

**`src/server/operator.rs` — cover the SSE wire branches (Minor).** Valid, fixed. `projects_task_id_only_when_the_event_is_correlated` asserts `taskId` present for a correlated `AgentReply` and `McpCallFailed` and absent when `task_id` is `None` (the existing test also now asserts absence, so the pre-#185 wire shape is pinned byte-for-byte). `projects_desk_task_completed_with_every_field` covers all four fields.

**`src/server/ops/write_test.rs` — untested timeline branches (Nitpick).** Valid, and worth more than a nitpick — taken. `task_timeline_scopes_approvals_to_the_run_window` covers `tool_failed` and the window-correlated `approval` arm, which as you note is the only heuristic correlation in the fold and therefore the one most likely to regress into leaking another run.

It asserts the window from **both** sides: an `ApprovalResolved` before the dispatch anchor and one after the window closed are both excluded, and only the one inside is admitted. It also asserts the operator identity (`u-1`) never reaches the wire, matching the SSE projection deny-by-default stance.

Full gate re-run on the updated branch: `cargo fmt --all -- --check`, `cargo clippy --all-targets -- -D warnings`, `cargo test` (766 passed, 0 failed), plus `cargo check --lib --tests --features openhuman` for the `run_task` changes — the default build does not compile `src/harness/`, so that last one is the only thing that checks them.

File: src/server/ops/tasks.rs [hunk 3]
```diff
@@ -107,10 +135,70 @@ async fn list_tasks(company: ScopedCompany) -> Result<Json<Vec<TaskCard>>, ApiEr
...
+
 async fn create_task(
     company: ScopedCompany,
     Json(body): Json<CreateTask>,
 ) -> Result<Json<TaskCard>, ApiError> {
+    if let Some(parent) = body.parent_task_id.as_deref() {
+        let board = company.runtime.tasks().list(company.id()).await?;
+        // `None` child: the card does not exist yet, so it cannot be in a cycle.
+        validate_parent(parent, None, &board)?;
+    }
```
[the remark this answers is not in the document]
Comment (M3gA-Mind) [reply src/server/ops/tasks.rs:206]: Fixed in 264502c — confirmed real, and it reproduces deterministically rather than being a narrow window.

I took the first option (hold one lock across list→validate→upsert) but **not** on `CompanyRuntime::serial`, deliberately. That lock is held for the duration of a cycle, which is a live agent turn — reusing it would park every board edit behind an LLM call, so dragging a card would block on whatever the agent happens to be doing. Instead there is a new per-company `CompanyRuntime::task_writes` mutex, only ever held across a couple of store round-trips.

Two extensions beyond the literal suggestion:

- **`patch_task` holds it for the whole handler**, not just the re-parent branch, which also makes the surrounding read-modify-write of the card's other fields lost-update free.
- **`delete_task` takes it too**, so a delete cannot land between a concurrent re-parent's existence check and its write — that is the dangling-edge case `validate_parent` exists to prevent, reached by a different route.

No deadlock with `upsert_task`: it does not take this lock, and the dispatch it may trigger is detached.

The regression test (`concurrent_reparents_cannot_race_a_cycle_onto_the_board`) fires both halves of `A → B` / `B → A` concurrently. Against the unguarded handlers it fails with `[200, 200]` and a full cycle persisted on the board; with the writes serialized, whichever request takes the lock second sees the first one's edge and is rejected — so it asserts *exactly* one success and a board that is still a forest, rather than "usually one".

I did not take the atomic-store-operation option: that means a new combined validate+upsert on the `TaskStore` trait and all four implementations, which is a much larger change than the race warrants, and single-node serialization is the same guarantee every other write path here has.

**Comment by M3gA-Mind** [comment]
Addressing the two outside-diff items from the latest review, since neither could be posted inline.

**`surface_mcp_failures` dropping the rest of the batch (`brain.rs` 436-465) — fixed in b04bfb5.** Confirmed real, and the reason it matters is slightly worse than "the batch aborts": `McpFailureQueue::drain` is a `mem::take`, so by the time the loop runs the queue is already empty and the batch exists only in that iterator. The `?` did not leave anything to retry from — the remaining failures were gone for good.

Worth noting it was also *quiet*. The call site in `run_task` already wraps the whole drain in `if let Err`, so this never aborted a cycle; it silently truncated the audit trail, which is the harder failure to notice.

Now per-item best-effort — a failed append logs its server and task id and the drain continues — matching the post-persist writes in `run_task` and `park_approval_requests`. The regression test wires an event log whose *first* append fails, queues three failures, and asserts all three still surface as error steps and that the two behind the failed write reach the journal. Against the old code it fails outright with `Store("journal unavailable")` propagating out of the drain.

**Stale duplicate doc comment on `run_task` (102-115) — taken in b04bfb5.** Two summary paragraphs were indeed stacked. For the record it was not from this PR: it arrived on `main` in #179 and I inherited it through the conflict resolution. Trivial and in a file this PR already touches, so I folded them into one rather than leaving it for someone else.

The one inline actionable (task-write serialization) is answered on its own thread, fixed in 264502c.

Verification for both commits: `cargo fmt --check` clean, default lane 770 passed / 0 failed, and `--features openhuman` 1068 passed / 1 failed. That single failure is `dispatched_card_with_an_origin_completes_to_done`, which is #179's own test and reproduces identically on pristine `upstream/main` — it calls `run_task` directly without the `pool.ensure` that `run_cycle` does, so the dispatch always errors to `backlog`. Not from this PR, and invisible to CI because the default lane does not build the `openhuman` feature.

File: src/server/ops/tasks.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jay-babu) [reply src/server/ops/tasks.rs:206]: Fixed in `c00fb9e7` without a lock, claim, attempt path, or shared S3 staging. Each market generation is encoded in-process, then publication deletes only canonical `_SUCCESS`, fails on any legacy `DeleteObjects.Errors`, writes all four intended source objects, verifies exact readback, creates `_SUCCESS` with `IfNoneMatch="*"`, and post-verifies stable marker/content/marker identity with bounded retries. Failed/smoke runs never mutate the prior snapshot. Forced two-writer interleaving passed 20 repetitions; the final marker always bound one complete generation and no batch/claim keys existed.
